### PR TITLE
Update nft module internal functions

### LINKED ISSUE #9004: Update NFT module to use internal functions to save data to stores
### Description

- The NFT module defines the createNFTEntry function which is responsible for creating a new NFT and ensuring that the NFT’s attributesArray array does not have duplicate modules
- The createNFTEntry function is never called. Instead, the nftStore.save method is called directly in many locations. In some locations, such as in the create method, the attributeArray uniqueness check is copy-pasted, while in others this uniqueness property is not enforced.
- In particular, two places may result in the NFT having duplicate module attributes
 When a foreign NFT is received (lisk-sdk/framework/src/modules/nft/cc_commands/cc_transfer.ts#L142-L145)
 When a foreign NFT is bounced (lisk-sdk/framework/src/modules/nft/cc_commands/cc_transfer.ts#L149-L152)
- Other instances
lisk-sdk/framework/src/modules/nft/cc_commands/cc_transfer.ts#L116 
lisk-sdk/framework/src/modules/nft/method.ts#L977
- Similarly createUserEntry function is not used [here](https://github.com/LiskHQ/lisk-sdk/blob/ed5649eb954c7c47e11eb2d2ea2b84b9336c4c4b/framework/src/modules/nft/method.ts#L327-L330) and [here](https://github.com/LiskHQ/lisk-sdk/blob/ed5649eb954c7c47e11eb2d2ea2b84b9336c4c4b/framework/src/modules/nft/module.ts#L290-L292)
-  Similarly createEscrowEntry function is not used [here](https://github.com/LiskHQ/lisk-sdk/blob/ed5649eb954c7c47e11eb2d2ea2b84b9336c4c4b/framework/src/modules/nft/module.ts#L288)

### Acceptance Criteria

- All places, including the ones mentioned above where internal functions can be used to save data to stores must be checked and updated
- Unit tests must be added to check there are no duplicate module attributes

### Additional Information
release/6.1.0

### PR DESCRIPTION
**Pull request by mosmartin**
### What was the problem?

This PR resolves #9004

### How was it solved?

- refactored the `createNFTEntry`, `createEscrowEntry` and `createUserEntry ` methods of the `InternalMethod` class
- added the `hasDuplicateModuleNames` method to the `InternalMethod` class
- refactored the `execute` method of the `CrossChainTransferCommand` class
- refactored the `create` method of the `NFTMethod` class
- refactored the `initGenesisState` of the `NFTModule` class

### How was it tested?

Added and updated existing tests

### CODE REVIEW

File: framework/src/modules/nft/module.ts [hunk 1]
```diff
@@ -273,24 +272,15 @@ export class NFTModule extends BaseInteroperableModule {
...
 
-			await nftStore.save(context, nftID, {
-				owner,
-				attributesArray,
-			});
+			await this._internalMethod.createNFTEntry(context, owner, nftID, attributesArray);
 
 			if (owner.length === LENGTH_CHAIN_ID) {
-				await escrowStore.set(context, escrowStore.getKey(owner, nftID), {});
+				await this._internalMethod.createEscrowEntry(context, owner, nftID);
```
[the remark this answers is not in the document]
Comment (mosmartin) [reply framework/src/modules/nft/module.ts:281]: Oh nice. I only added the `GenesisBlockExecuteContext` as I wasn't sure if there was an alternative. Lemme try it. Thanks @has5aan for pointing this out.

File: framework/src/modules/nft/cc_commands/cc_transfer.ts [hunk 2]
```diff
@@ -105,24 +105,24 @@ export class CrossChainTransferCommand extends BaseCCCommand {
...
+			if (status !== CCM_STATUS_CODE_OK) {
 				recipientAddress = senderAddress;
-				storeData.owner = recipientAddress;
-				await nftStore.save(getMethodContext(), nftID, storeData);
-				await this._internalMethod.createUserEntry(getMethodContext(), recipientAddress, nftID);
-				await escrowStore.del(getMethodContext(), escrowStore.getKey(sendingChainID, nftID));
 			}
+
+			storeData.owner = recipientAddress;
+			await nftStore.save(getMethodContext(), nftID, storeData);
```
Comment (has5aan) [inline framework/src/modules/nft/cc_commands/cc_transfer.ts:123]: > In particular, two places may result in the NFT having duplicate module attributes
When a foreign NFT is received (lisk-sdk/framework/src/modules/nft/cc_commands/cc_transfer.ts#L142-L145)
When a foreign NFT is bounced (lisk-sdk/framework/src/modules/nft/cc_commands/cc_transfer.ts#L149-L152)

As per this point in the issue; to avoid duplicate attributes, perhaps createNFTEntry is to be called here as well?! 🤔  [LIP's definition of the method](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0052.md#execution-2)

[the remark this answers is not in the document]
Comment (has5aan) [reply framework/src/modules/nft/cc_commands/cc_transfer.ts:123]: Similar to this, the second point in the same section on issue description;

> Other instances
lisk-sdk/framework/src/modules/nft/cc_commands/cc_transfer.ts#L116
lisk-sdk/framework/src/modules/nft/method.ts#L977

The second point is about `NFTMethod.recover`; here again the validation of duplicate module names gets skipped as save is called on the NFTStore directly. However, [LIP](https://github.com/LiskHQ/lips/blob/1110f660010f91331e4178e392a22349f5c167f1/proposals/lip-0052.md?plain=1#L1999-L2004) doesn't explicitly points that out.

**Review by has5aan — APPROVED** [review]

**Review by has5aan — APPROVED** [review]

File: framework/src/modules/nft/cc_commands/cc_transfer.ts [hunk 3]
```diff
@@ -104,24 +104,29 @@ export class CrossChainTransferCommand extends BaseCCCommand {
...
 					storedAttributes,
 					receivedAttributes,
 				);
-				await nftStore.save(getMethodContext(), nftID, storeData);
-				await this._internalMethod.createUserEntry(getMethodContext(), recipientAddress, nftID);
-				await escrowStore.del(getMethodContext(), escrowStore.getKey(sendingChainID, nftID));
-			} else {
+			}
+
+			if (status !== CCM_STATUS_CODE_OK) {
```
Comment (Incede) [inline framework/src/modules/nft/cc_commands/cc_transfer.ts:117]: Why replace `if else` with slower `if if` ?

File: framework/src/modules/nft/cc_commands/cc_transfer.ts [hunk 4]
```diff
@@ -104,24 +104,29 @@ export class CrossChainTransferCommand extends BaseCCCommand {
...
 				recipientAddress = senderAddress;
-				storeData.owner = recipientAddress;
-				await nftStore.save(getMethodContext(), nftID, storeData);
-				await this._internalMethod.createUserEntry(getMethodContext(), recipientAddress, nftID);
-				await escrowStore.del(getMethodContext(), escrowStore.getKey(sendingChainID, nftID));
 			}
+
+			storeData.owner = recipientAddress;
```
Comment (Incede) [inline framework/src/modules/nft/cc_commands/cc_transfer.ts:121]: I believe this logic has no purpose now and can be removed

File: framework/src/modules/nft/method.ts [hunk 5]
```diff
@@ -296,13 +296,7 @@ export class NFTMethod extends BaseMethod {
...
 		attributesArray: NFTAttributes[],
 	): Promise<void> {
-		const moduleNames = [];
-		for (const item of attributesArray) {
-			moduleNames.push(item.module);
-		}
-		if (new Set(moduleNames).size !== attributesArray.length) {
-			throw new Error('Invalid attributes array provided');
-		}
+		this._internalMethod.hasDuplicateModuleNames(attributesArray);
```
Comment (Incede) [inline framework/src/modules/nft/method.ts:299]: This check is redundant.

**Review by Incede — CHANGES_REQUESTED** [review]
I think it'd be nice to add unit test to check if NFT has duplicate module attributes for these 2 scenarios :
1. When a foreign NFT is received (lisk-sdk/framework/src/modules/nft/cc_commands/cc_transfer.ts#L142-L145)
2. When a foreign NFT is bounced (lisk-sdk/framework/src/modules/nft/cc_commands/cc_transfer.ts#L149-L152)

File: framework/src/modules/nft/cc_commands/cc_transfer.ts (hunk 3 above)
Comment (mosmartin) [reply framework/src/modules/nft/cc_commands/cc_transfer.ts:117]: 1. The `getMethodContext()` function is called multiple times. If this function is computationally expensive, it might be beneficial to call it once and store the result in a variable.

2. The `escrowStore.getKey(sendingChainID, nftID)` is also called twice. This could be stored in a variable to avoid duplicate computation.

Comment (Incede) [reply framework/src/modules/nft/cc_commands/cc_transfer.ts:117]: sure, it'd be nice to also update the other places which can be improved. but was there a specific reason to change from `if else` to `if if` ?

Comment (mosmartin) [reply framework/src/modules/nft/cc_commands/cc_transfer.ts:117]: Main reason was to remove the duplication of the calls below

```ts
await nftStore.save(getMethodContext(), nftID, storeData);
await this._internalMethod.createUserEntry(getMethodContext(), recipientAddress, nftID);
await escrowStore.del(getMethodContext(), escrowStore.getKey(sendingChainID, nftID));

Comment (mosmartin) [reply framework/src/modules/nft/cc_commands/cc_transfer.ts:117]: If we optimize this code for performance, 

```ts
if (status === CCM_STATUS_CODE_OK) {
  const { attributesArray: storedAttributes } = storeData;
  storeData.attributesArray = this._internalMethod.getNewAttributes(
    nftID,
    storedAttributes,
    receivedAttributes,
  );
} else {
 recipientAddress = senderAddress;
}

const { attributesArray } = storeData;
const createNFTEntryPromise = this._internalMethod.createNFTEntry(
  getMethodContext(),
  recipientAddress,
  nftID,
  attributesArray,
);
const createUserEntryPromise = this._internalMethod.createUserEntry(
  getMethodContext(),
  recipientAddress,
  nftID,
);
const delPromise = escrowStore.del(
  getMethodContext(),
  escrowStore.getKey(sendingChainID, nftID),
);

await Promise.all([createNFTEntryPromise, createUserEntryPromise, delPromise]);
```

would be a better option as it can significantly reduce the overall execution time of a group of asynchronous operations when compared to executing them sequentially

**Comment by mosmartin** [comment]
> I think it'd be nice to add unit test to check if NFT has duplicate module attributes for these 2 scenarios :
> 
> 1. When a foreign NFT is received (lisk-sdk/framework/src/modules/nft/cc_commands/cc_transfer.ts#L142-L145)
> 2. When a foreign NFT is bounced (lisk-sdk/framework/src/modules/nft/cc_commands/cc_transfer.ts#L149-L152)

@Incede for the first test, isn't it covered by 
`it('should reject and emit unsuccessful ccm transfer event if nft chain id does not equal own chain id and nft is not supported', async () => {...}`?

**Comment by Incede** [comment]
> > I think it'd be nice to add unit test to check if NFT has duplicate module attributes for these 2 scenarios :
> > 
> > 1. When a foreign NFT is received (lisk-sdk/framework/src/modules/nft/cc_commands/cc_transfer.ts#L142-L145)
> > 2. When a foreign NFT is bounced (lisk-sdk/framework/src/modules/nft/cc_commands/cc_transfer.ts#L149-L152)
> 
> @Incede for the first test, isn't it covered by `it('should reject and emit unsuccessful ccm transfer event if nft chain id does not equal own chain id and nft is not supported', async () => {...}`?

I don't think the test case you mention is related to any of the 2 scenarios :
1. foreign NFT is received
2. foreign NFT is received

plus my suggestion is to check if NFT has duplicate module attributes for these 2 scenarios which isn't being currently checked, since the original issue was primarily about allowing "duplicate module attributes" and if the suggested test was already there the issue would have been caught before. Although I understand that the function `createNFTEntry` is already tested for duplicate module attributes I thought it'd be nice to test it again here for the sake of this issue.

**Review by Incede — APPROVED** [review]
